src/mon/LogMonitor.cc

```diff
@@ -267,134 +297,301 @@ void LogMonitor::update_from_paxos(bool *need_bootstrap)
-
-      string channel = le.channel;
-      if (channel.empty()) // keep retrocompatibility
-        channel = CLOG_CHANNEL_CLUSTER;
-
-      if (g_conf().get_val<bool>("mon_cluster_log_to_stderr")) {
-	cerr << channel << " " << le << std::endl;
+    __u8 struct_v;
+    decode(struct_v, p);
+    if (struct_v == 1) {
+      // legacy pre-quincy commits
+      while (!p.end()) {
+	LogEntry le;
+	le.decode(p);
+	dout(7) << "update_from_paxos applying incremental log "
+		<< summary.version+1 <<  " " << le << dendl;
+	summary.add_legacy(le);
       }
-
-      if (channels.do_log_to_syslog(channel)) {
-        string level = channels.get_level(channel);
-        string facility = channels.get_facility(channel);
-        if (level.empty() || facility.empty()) {
-          derr << __func__ << " unable to log to syslog -- level or facility"
-               << " not defined (level: " << level << ", facility: "
-               << facility << ")" << dendl;
-          continue;
-        }
-        le.log_to_syslog(channels.get_level(channel),
-                         channels.get_facility(channel));
-      }
-
-      if (channels.do_log_to_graylog(channel)) {
-	ceph::logging::Graylog::Ref graylog = channels.get_graylog(channel);
-	if (graylog) {
-	  graylog->log_log_entry(&le);
+    } else {
+      uint32_t num;
+      decode(num, p);
+      while (num--) {
+	LogEntry le;
+	le.decode(p);
+	dout(7) << "update_from_paxos applying incremental log "
+		<< summary.version+1 <<  " " << le << dendl;
+	summary.recent_keys.insert(le.key());
+	summary.channel_info[le.channel].second++;
+	// we may have logged past the (persisted) summary in a prior quorum
+	if (version > external_log_to) {
+	  log_external(le);
 	}
-	dout(7) << "graylog: " << channel << " " << graylog
-		<< " host:" << channels.log_to_graylog_host << dendl;
       }
-
-      if (channels.do_log_to_journald(channel)) {
-        auto &journald = channels.get_journald();
-        journald.log_log_entry(le);
-        dout(7) << "journald: " << channel << dendl;
-      }
-
-      if (g_conf()->mon_cluster_log_to_file) {
-	string log_file = channels.get_log_file(channel);
-	dout(20) << __func__ << " logging for channel '" << channel
-		 << "' to file '" << log_file << "'" << dendl;
-
-	if (!log_file.empty()) {
-	  string log_file_level = channels.get_log_file_level(channel);
-	  if (log_file_level.empty()) {
-	    dout(1) << __func__ << " warning: log file level not defined for"
-		    << " channel '" << channel << "' yet a log file is --"
-		    << " will assume lowest level possible" << dendl;
-	  }
-
-	  int min = string_to_syslog_level(log_file_level);
-	  int l = clog_type_to_syslog_level(le.prio);
-	  if (l <= min) {
-	    stringstream ss;
-	    ss << le << "\n";
-	    // init entry if DNE
-	    bufferlist &blog = channel_blog[channel];
-	    blog.append(ss.str());
-	  }
-	}
+      map<string,version_t> prune_channels_to;
+      for (auto& p : prune_channels_to) {
```

Comment:
shall we decode the rest stuff in `p` before iterating through `prune_channels_to`?

Reply:
and better off using structured binding like,
```c++
for (auto& [channel, prune_to] : prune_channels_to) {
  // ...
  summary.channel_info[channel].first = prune_to;
}
```